notificationworkerlambda/src/main/scala/com/gu/notifications/worker/SenderRequestHandler.scala

```diff
@@ -105,4 +130,25 @@ trait SenderRequestHandler[C <: DeliveryClient] extends Logging {
 
     logEndOfInvocation(sqsMessageBatchSize, totalTokensProcessed, startTime, registry)
   }
+
+  def handleRequest(input: InputStream, output: OutputStream, context: Context): Unit = {
+    val json: JsValue = Json.parse(input);
+    val sqsQueue = (json \ "resources" \ 0).get.as[String]
+    val batchSizeToUse = (json \ "detail" \ "batchsize").toOption.map(_.as[Int]).getOrElse(batchSize)
+
+    val sqsClient = AmazonSQSClientBuilder.standard()
+      .withCredentials(Aws.credentialsProvider)
+      .withRegion(Regions.EU_WEST_1)
+      .build()
+
+    val receiveMessages = sqsClient.receiveMessage(new ReceiveMessageRequest()
+      .withQueueUrl(sqsQueue)
+      .withMaxNumberOfMessages(batchSizeToUse)
+      .withWaitTimeSeconds(1)
+      .withAttributeNames("SentTimestamp"))
+      .getMessages.asScala.toList
+
+    handleChunkTokensInMessages(receiveMessages.map(SqsMessage.fromSqsApiMessage(_)))
+    receiveMessages.foreach(msg => sqsClient.deleteMessage(sqsQueue, msg.getReceiptHandle()))   
```

Comment:
Am I understanding it correctly: given we've successfully handled the chunked tokens, we delete the messages from the queue? And this functionality is no different from what we currently do: if any message in the batch throws an unhandled error, all messages from the batch will be retried?

Reply:
Yes, that's right.  

When we finish processing the chunked tokens, we will delete the messages at the end.  

If any exceptions are thrown and unhandled, the lambda will terminate, leaving the messages in the queue.  In this case, the lambda returns an error and the EventBridge will [retry calling the function again](https://docs.aws.amazon.com/eventbridge/latest/userguide/eb-rule-dlq.html) (the maximum retry count is set to 5 in our CDK).

I think if unhandled exceptions happen, they probably results from JVM runtime error or getting AWS client instance.  It looks less likely to come from the processing of individual messages because we handle exceptions that may happen from APNS / Firebase API and put them in the output (`Either[DeliveryException, DeliverySuccess]`) of the stream.